### PR TITLE
machines: create VM with installation source PXE

### PR DESCRIPTION
**Pull request by KKoukiou**
This will have merge conflicts with https://github.com/cockpit-project/cockpit/pull/11206
I would like to give priority to the other one since it's simpler and will have less review iterations.

create-vm-pxe-select
create-vm-pxe-source-list
create-vm-network-boot-macvtap
pxe-boot-screen

### CODE REVIEW

**Comment by KKoukiou** [comment]
@andreasn or @garrett this is an quite early stage implementation of this feature, but I thought that you might like to give some design feedback while l am working on the tests?

**Comment by KKoukiou** [comment]
@Antique I think you would like to review this.

**Comment by andreasn** [comment]
It's funny with the "in most configurations..." error message, because I get it in virt-manager _all the time_, but it Always Works. :)
What's the story behind this notification? Is there anywhere I can read more about it?

**Comment by KKoukiou** [comment]
@martinpitt This is failing on some licencies compatibilty :/ NPM sax module that am adding in the dependencies is coming from that xmljs  that I am using from the xmljs module (which we decided to not import)  but instead 'copy' the parts that we need.

Is this really a problem? If so I will try to go with the solution to parse the domparser element directly instead of building this js object from xml.

**Comment by martinpitt** [comment]
KeyError: 'sax has license ISC which is not in ./tools/debian/copyright.template'

@KKoukiou : ISC is fine, it just needs to be added to tools/debian/copyright.template . You can copy the license text from https://spdx.org/licenses/ISC.html , or most probably it's shipped within `node_modules/sax/`, too.

**Comment by andreasn** [comment]
Looks good in general!
The orange text on white is a little hard to read, so maybe an /!\ icon + black text could work instead. Or maybe something in a little info bubble? Not sure how to sort that best.
Does the host-to-client communication affect the PXE-booting itself?

**Comment by KKoukiou** [comment]
> Looks good in general!
> The orange text on white is a little hard to read, so maybe an /!\ icon + black text could work instead. 

That yellow is coming from bootstrap warning class in HelpBlock, I didn't overwrite it or anything. But I was thinking as well if tooltip if preferred. But it would be weird that tooltip would appear and disappear when filling the form.

> Or maybe something in a little info bubble? Not sure how to sort that best.
> Does the host-to-client communication affect the PXE-booting itself?

No, the host to guest communication does not have any impact on the PXE booting at all.

File: pkg/machines/scripts/install_machine.sh [hunk 1]
```diff
@@ -47,7 +48,9 @@ else
     GRAPHICS_PARAM="$CREATE_OPTIONS_RESULT"
 fi
 
-if [ "${SOURCE#/}" != "$SOURCE" ] && [ -f "${SOURCE}" ]; then
+if [ "$SOURCE_TYPE" == "pxe" ]; then
+        LOCATION_PARAM="--pxe --network $SOURCE"
```
Comment (Antique) [inline pkg/machines/scripts/install_machine.sh]: Before this change we don't specify `--network` when executing `virt-install` command which means that the created/installed guest will have the `default` NAT network.  With this change when the guest is installed using `PXE` method it will not have the `default` network like for all other installation methods but the selected network in the creation dialog.

The reason why I'm mentioning it is that for filesystem or URL installation source type the installation source data are used only for installation and are removed after the installation is done, but in case of PXE the source is not used only for installation but it configures the main and only network for the guest once the installation is done.

The question is do we want to have it the same way as virt-manager that in the creation dialog you need to select the installation source or will we leave the guest interface configuration up to the user?  In both cases we should improve guest interface configuration because right now you can only edit existing interface but you cannot add/remove interfaces.

**Comment by Antique** [comment]
> It's funny with the "in most configurations..." error message, because I get it in virt-manager _all the time_, but it Always Works. :)
> What's the story behind this notification? Is there anywhere I can read more about it?

Does the guest - host communication really work?  Some explanation is here [https://wiki.libvirt.org/page/TroubleshootMacvtapHostFail](https://wiki.libvirt.org/page/TroubleshootMacvtapHostFail).

**Comment by KKoukiou** [comment]
I need to allow only libvirt dbus provider for this feature, virsh provider does not have info about node devices.

File: pkg/machines/scripts/install_machine.sh (hunk 1 above)
Comment (KKoukiou) [reply pkg/machines/scripts/install_machine.sh]: Good point, thanks for raising these issues/thoughts. So, I think we should definitely allow the iser to select the installation source, whatever that is. Otherwise the user will never be able to start immediately a VM from PXE if his default network is not configured like that. Indeed the NICs dialog needs enhancements as well. But I don't think we should rely on that the user will create the VM with PXE install mode and then go to the Interfaces tab to configure them before starting the VM.
And of course we need to find a more decent to offer to users who need advanced configurations, since this applies to other parameters apart from pxe booting.

File: pkg/machines/xml-helpers/array-helper.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// This is copied from https://github.com/nashwaan/xml-js/blob/master/lib/array-helper.js
+
+export function isArray(value) {
+    if (Array.isArray) {
+        return Array.isArray(value);
+    }
+    // fallback for older browsers like  IE 8
+    return Object.prototype.toString.call(value) === '[object Array]';
```
Comment (martinpitt) [inline pkg/machines/xml-helpers/array-helper.js]: We don't support these, so the whole helper function can be dropped. We can rely on `Array.isArray()`.

File: pkg/machines/xml-helpers/options-helper.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+export function copyOptions(options) {
+    var key; var copy = {};
+    for (key in options) {
+        if (options.hasOwnProperty(key)) {
+            copy[key] = options[key];
+        }
+    }
+    return copy;
+}
```
Comment (martinpitt) [inline pkg/machines/xml-helpers/options-helper.js]: How is `copyOptions(o)` different from `{...o}` (or `Object.assign({}, o)`)?

File: pkg/machines/xml-helpers/xml2js.js [hunk 4]
```diff
@@ -0,0 +1,363 @@
+// This is copied from https://github.com/nashwaan/xml-js/blob/master/lib/xml2js.js
+
+import * as sax from 'sax';
```
Comment (martinpitt) [inline pkg/machines/xml-helpers/xml2js.js]: I suppose there was some misunderstanding. If we have to copy that much code, then let's just use xml-js; having verbatim code copies is worse. It's not literally the dependency declaration in package.json that is the problem, but having to add (externally or internally) a lot of new code.

My question back then was about whether we really need to do this XML → JS → XML conversion dance in the first place, as it's lossy and at least structurally not really necessary. If manipulating XML is too unwieldy, then it seems better to me to introduce some helper functions for *that*.

File: pkg/machines/libvirt-dbus.js [hunk 5]
```diff
@@ -460,6 +464,21 @@ LIBVIRT_DBUS_PROVIDER = {
...
+    GET_ALL_NODE_DEVICES({
+        connectionName,
+    }) {
+        return dispatch => {
+            call(connectionName, '/org/libvirt/QEMU', 'org.libvirt.Connect', 'ListNodeDevices', [0], TIMEOUT)
+                    .then(objPaths => {
+                        // We can't use Promise.all() here until cockpit is able to dispatch es2015 promises
+                        // https://github.com/cockpit-project/cockpit/issues/10956
+                        // eslint-disable-next-line cockpit/no-cockpit-all
+                        return cockpit.all(objPaths[0].map(path => dispatch(getNodeDevice({ connectionName, id:path }))));
```
Comment (martinpitt) [inline pkg/machines/libvirt-dbus.js]: OOI, where do Cockpit promises come into play here? redux certainly uses ES6 Promise?

File: pkg/machines/libvirt-dbus.js [hunk 6]
```diff
@@ -541,6 +561,32 @@ LIBVIRT_DBUS_PROVIDER = {
...
+    GET_NODE_DEVICE({
+        id: objPath,
+        connectionName,
+    }) {
+        let props = {};
+
+        return dispatch => {
+            call(connectionName, objPath, 'org.libvirt.NodeDevice', 'GetXMLDesc', [0], TIMEOUT)
+                    .then(deviceXml => {
+                        let result = xml2js(deviceXml, { compact: true });
```
Comment (martinpitt) [inline pkg/machines/libvirt-dbus.js]: Wrt. above, this is the bit that I'm cautious about: Is this really reversible, or can it be lossy?

File: pkg/machines/libvirt-dbus.js [hunk 7]
```diff
@@ -541,6 +561,32 @@ LIBVIRT_DBUS_PROVIDER = {
...
+
+                        props.connectionName = connectionName;
+                        props.xml = result;
```
Comment (martinpitt) [inline pkg/machines/libvirt-dbus.js]: That's confusing, as this is now a JS object, not XML. (But of course, let's please try and keep it as XML; a string, or possibly a [DOMParser](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser) object)

File: pkg/machines/libvirt-dbus.js [hunk 8]
```diff
@@ -555,6 +555,13 @@ LIBVIRT_DBUS_PROVIDER = {
                         props.id = objPath;
                         props.connectionName = connectionName;
 
+                        return call(connectionName, objPath, 'org.libvirt.Network', 'GetXMLDesc', [0], TIMEOUT);
+                    })
+                    .then(networkXml => {
+                        let result = xml2js(networkXml, { compact: true });
+
+                        props.xml = result;
```
Comment (martinpitt) [inline pkg/machines/libvirt-dbus.js]: It feels like this commit ("add network resource XML to the store object") should be part of another commit that actually makes use of it.

File: test/verify/machineslib.py [hunk 9]
```diff
@@ -1104,14 +1131,41 @@ def installWithErrorTest(dialog):
...
                                                    os_vendor=config.APPLE_VENDOR,
                                                    os_name=config.MACOS_X_LEOPARD,
                                                    connection='session'))
 
+        # test PXE Source
+        if self.provider == "libvirt-dbus":
+            self.machine.execute("virsh net-destroy default && virsh net-undefine default")
+
+            # Disable selinux because it makes TFTP directory inaccesible and we don't want sophisticated configuration for tests
+            self.machine.execute("setenforce 0")
```
Comment (martinpitt) [inline test/verify/machineslib.py]: This needs to be guarded, as `setenforce` doesn't exist on the Debian based OSes.

File: test/verify/machineslib.py [hunk 10]
```diff
@@ -1104,14 +1131,41 @@ def installWithErrorTest(dialog):
...
+            self.browser.wait_in_text("body", "Virtual Machines")
+
+            runner.tryCreate(TestMachines.VmDialog(self, "subVmTestCreate16", sourceType='pxe',
+                                                   location="Virtual Network pxe-nat: NAT",
+                                                   memory_size=1, memory_size_unit='GiB',
+                                                   storage_size=100, storage_size_unit='MiB',
+                                                   os_vendor=config.NOVELL_VENDOR,
+                                                   os_name=config.NOVELL_NETWARE_6,
+                                                   start_vm=True))
+            self.browser.wait_in_text("#vm-subVmTestCreate16-state", "running")
```
Comment (martinpitt) [inline test/verify/machineslib.py]: Do we have a way to validate that the PXE boot actually worked? "Running" just means the (virt-)phsyical machine state, so it could just hang in the BIOS/EFI with no working boot device?

File: pkg/machines/xml-helpers/xml2js.js (hunk 4 above)
Comment (KKoukiou) [reply pkg/machines/xml-helpers/xml2js.js]: We do need to change attributes in existing domains XML. That's why I thought that having the XMLs as object will result in cleaner code. But of course it's not the only way to do it.

So, I will rewrite this PR and instead of having all these xmljs helpers I will parse directly the DomParser element and save it with the Domain description object. 
I trust you that it might not be that bad in the end to just directly parse the DOMPrser.

**Comment by KKoukiou** [comment]
This PR needs to be rewritten to parse the DOMParser directly instead of introducing these xmljs functions.

File: pkg/machines/libvirt-dbus.js (hunk 7 above)
Comment (KKoukiou) [reply pkg/machines/libvirt-dbus.js]: I will rewrite the PR in this direction, np :)

File: pkg/machines/xml-helpers/array-helper.js (hunk 2 above)
Comment (KKoukiou) [reply pkg/machines/xml-helpers/array-helper.js]: I removed completely the xml helpers.. In the end I think just parsing the DOMParser will do just fine. Sorry for the noise with xmljs.

File: pkg/machines/xml-helpers/options-helper.js (hunk 3 above)
Comment (KKoukiou) [reply pkg/machines/xml-helpers/options-helper.js]: Removed xml-helpers.

File: pkg/machines/libvirt-dbus.js (hunk 6 above)
Comment (KKoukiou) [reply pkg/machines/libvirt-dbus.js]: Removed xml2hs helper. Now if we need to change the existing XML we 'll use the domParser object itself.

File: pkg/machines/libvirt-dbus.js (hunk 8 above)
Comment (KKoukiou) [reply pkg/machines/libvirt-dbus.js]: Makes sense. Simon introduced network details already in his Networks PR that was merged last week, so with rebase I will have them ready for use.

File: test/verify/machineslib.py (hunk 9 above)
Comment (KKoukiou) [reply test/verify/machineslib.py]: Right, thanks!

File: test/verify/machineslib.py (hunk 10 above)
Comment (KKoukiou) [reply test/verify/machineslib.py]: I was struggling with this to be honest. I am aware that it's not 'fully' tested like this. I will try to find a more fine grained way to test this.

**Comment by martinpitt** [comment]
This looks really nice code-wise! I tested this, and the completion is now much more reactive. However, I have the feeling that it  doesn't throttle at all. The completion result gets updated on every key stroke even on fast typing.

I bumped the throttle timeout to 3s (value 3000), and it's still the same.

**Comment by martinpitt** [comment]
I force-pushed a fix to the grepping of the serial console, to filter out the ANSI control character crap. This  works now, but the test fails later on:
```
  File "/home/martin/upstream/c2/test/verify/machineslib.py", line 1239, in testCreate
    dialog = {delete: True, name: 'subVmTestCreate18'}
NameError: name 'delete' is not defined
```

File: test/verify/machineslib.py [hunk 11]
```diff
@@ -22,6 +22,7 @@
 import subprocess
 import time
 import unittest
+import xml.etree.ElementTree as ET
```
Comment (KKoukiou) [inline test/verify/machineslib.py]: I can use sed instead of ET as well. I don't really have a preference here.

**Comment by KKoukiou** [comment]
Pff, [debian-testing](https://fedorapeople.org/groups/cockpit/logs/pull-11179-20190306-161818-f8d73973-verify-debian-testing/log.html#251) complaining that it can't allocate 500M, will reduce, we don't need the memory anyway

File: test/common/testlib.py [hunk 12]
```diff
@@ -910,6 +910,9 @@ def check_journal_messages(self, machine=None):
             self.allowed_messages.append('audit: type=1400 audit(.*): avc:  denied  { mount } for .* comm="find" .*pcp_pmlogger_t.*')
             # HACK: https://bugzilla.redhat.com/show_bug.cgi?id=1662866
             self.allowed_messages.append('audit: type=1400 audit(.*): avc:  denied  { execute } for .* comm="which" .*pcp_pmlogger_t.*')
+            # HACK: https://bugzilla.redhat.com/show_bug.cgi?id=1406979 - the bug is for open operation on a file of the proc directory so it should be identical
```
Comment (martinpitt) [inline test/common/testlib.py]: This is an old bug from Fedora 25 in 2017 which got fixed long ago. Does this need to be reopened or a new one?

File: test/verify/machineslib.py (hunk 11 above)
Comment (martinpitt) [reply test/verify/machineslib.py]: Up to you. I find XML hard to manipulate with sed, but if it's just a small local change and not a structural one, you might get away with it. :wink:

File: test/common/testlib.py (hunk 12 above)
Comment (KKoukiou) [reply test/common/testlib.py]: In the BZ which is indeed closed and old It's mentioned there that the fix it's to dontaudit these accesses.
So the warning is supposed to remain there unless we will have AVCRules specifying to don't audit these. Or are these rules builtin to the selinux package?
/me needs to read selinux docs

Comment (KKoukiou) [reply test/common/testlib.py]: Ok, so after a bit reading I found out what they meant in the BZ. Solution was so manually configure auditd to not audit these messages. So just ignoring these logs here sounds reasonable to me. It's same like configuring auditd to not print them at all.

Comment (martinpitt) [reply test/common/testlib.py]: Ack, makes sense; the comment could clarify this a bit. Thanks!

Comment (KKoukiou) [reply test/common/testlib.py]: ok!

Comment (KKoukiou) [reply test/common/testlib.py]: I added a comment and added to the https://github.com/cockpit-project/cockpit/pull/11272#issuecomment-470498479 since it has exactly the same problem.

I guess that one will be merged faster to I will rebase this one.

File: pkg/machines/components/create-vm-dialog/pxe-helpers.js [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+
+/**
+ * Filters an array of node devices returning only devices of specific connection and capability.
+ * @param {array} nodeDevices - An array of object containing NodeDevices.
+ * @param {string} connectionName - The libvirt connection name.
+ * @param {string} type - The capability type, ex 'net'.
+ */
+function getNodeDevicesOfType(nodeDevices, connectionName, type) {
+    return nodeDevices.filter(nodeDevice => {
+        const capabilityType = nodeDevice.capability.type;
```
Comment (martinpitt) [inline pkg/machines/components/create-vm-dialog/pxe-helpers.js]: This is unnecessary boilerplate, the const is being used exactly once. (But matter of preference, I figure).

File: pkg/machines/components/create-vm-dialog/pxe-helpers.js [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+        } else if (mode == 'route') {
+            if (dev)
+                return cockpit.format('Route to $0', dev);
+            else
+                return 'Routed Network';
+        } else {
+            if (dev)
+                return cockpit.format('$0 to $1', mode, dev);
+            else
+                return cockpit.format('$0 network', mode.toUpperCase());
```
Comment (martinpitt) [inline pkg/machines/components/create-vm-dialog/pxe-helpers.js]: The strings here need i18n.

File: pkg/machines/components/create-vm-dialog/pxe-helpers.js [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+        return true;
+    }
+
+    let ips = virtualNetwork.ip;
+
+    for (let i = 0; i < ips.length; i++)
+        if (ips[i].dhcp.bootp)
+            return true;
+
+    return false;
```
Comment (martinpitt) [inline pkg/machines/components/create-vm-dialog/pxe-helpers.js]: This could be simplified with `return virtualNetwork.ip.find(ip => ip.dhcp.bootp)` (prepend a `!!` if returning a truthy value or `undefined` instead of `true`/`false` is not acceptable)

File: pkg/machines/components/create-vm-dialog/pxe-helpers.js [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+
+    /* Do not show to the user the libvirt virbrX-nic devices since these are
+     * supposed to be managed through Virtual Networks
+     */
+    let libvirtVirBridges = getLibvirtNetworkBridges(virtualNetworks);
+    netNodeDevices = netNodeDevices.filter(netNodeDevice => {
+        for (let i in libvirtVirBridges) {
+            if (netNodeDevice.capability.interface == (libvirtVirBridges[i] + '-nic'))
+                return false;
+        }
```
Comment (martinpitt) [inline pkg/machines/components/create-vm-dialog/pxe-helpers.js]: This is a quadratic loop again, let's just hope that users don't have too many devices and bridges :-)

This could be optimized a bit by checking in advance whether netNodeDevice's interface ends with `-nic`.

File: pkg/machines/components/create-vm-dialog/pxe-helpers.js [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+    const libvirtVirBridges = getLibvirtNetworkBridges(virtualNetworks);
+    let netNodeDevices = getNodeDevicesOfType(nodeDevices, connectionName, 'net');
+
+    netNodeDevices = netNodeDevices.filter(netNodeDevice => {
+        for (let i in libvirtVirBridges) {
+            if (netNodeDevice.capability.interface == (libvirtVirBridges[i] + '-nic'))
+                return false;
+        }
+        return true;
+    });
```
Comment (martinpitt) [inline pkg/machines/components/create-vm-dialog/pxe-helpers.js]: This is almost the same code as above, can you please factorize it out into a helper?

File: pkg/machines/components/create-vm-dialog/pxe-helpers.js [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+        </Select.SelectEntry>
+    );
+
+    let netNodeDevicesRows = netNodeDevices.map(netNodeDevice => {
+        let iface = netNodeDevice.capability.interface;
+
+        return (
+            <Select.SelectEntry
+                data={cockpit.format('type=direct,source=$0', iface)}
+                key={cockpit.format('type=direct,source=$0', iface)}>
```
Comment (martinpitt) [inline pkg/machines/components/create-vm-dialog/pxe-helpers.js]: Please move the `cockpit.format` into the assignment of `iface` to avoid doing it twice.

File: test/verify/machineslib.py [hunk 19]
```diff
@@ -1136,6 +1163,92 @@ def installWithErrorTest(dialog):
                                          start_vm=True))
         # End of tests for import existing disk as installation option
 
+        # test PXE Source
+        if self.provider == "libvirt-dbus":
+            self.machine.execute("virsh net-destroy default && virsh net-undefine default")
+
+            if not ('debian' in self.machine.image or 'ubuntu' in self.machine.image):
+                # Disable selinux because it makes TFTP directory inaccesible and we don't want sophisticated configuration for tests
+                self.machine.execute("setenforce 0")
```
Comment (martinpitt) [inline test/verify/machineslib.py]: If this works, it's ok. But it might be more robust to call `if type setenforce >/dev/null 2>&1; then setenforce 0; fi`.

File: pkg/machines/components/create-vm-dialog/pxe-helpers.js (hunk 16 above)
Comment (KKoukiou) [reply pkg/machines/components/create-vm-dialog/pxe-helpers.js]: Optimization mentioned in second line done. For time complexity I think we should not worry too much even for 100 net devices it will get a O(n3) -> 100^3 which is still taking less than a second. I tend to count approximately 10^7 operations as 1 second.

File: pkg/machines/components/create-vm-dialog/pxe-helpers.js (hunk 17 above)
Comment (KKoukiou) [reply pkg/machines/components/create-vm-dialog/pxe-helpers.js]: done

File: pkg/machines/components/create-vm-dialog/pxe-helpers.js (hunk 18 above)
Comment (KKoukiou) [reply pkg/machines/components/create-vm-dialog/pxe-helpers.js]: done

File: test/verify/machineslib.py (hunk 19 above)
Comment (KKoukiou) [reply test/verify/machineslib.py]: I replaced it with your script.

**Comment by KKoukiou** [comment]
Looks like there is still two failures:
rhel-8 selinux issues in journal messages.
debian and rhel-7-6 the previous test with the imported disk failed because VM failed to get deleted.

**Comment by martinpitt** [comment]
@KKoukiou : PR #11354 has the fix for the qemu SELinux message.

File: pkg/machines/components/create-vm-dialog/pxe-helpers.js [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+
+    if (forward) {
+        mode = forward.mode;
+        dev = forward.interface && forward.interface.dev;
+    }
+
+    if (mode || dev) {
+        if (!mode || mode == 'nat') {
+            if (dev)
+                return cockpit.format('NAT to $0', dev);
```
Comment (martinpitt) [inline pkg/machines/components/create-vm-dialog/pxe-helpers.js]: One more missing i18n.

File: pkg/machines/components/create-vm-dialog/pxe-helpers.js [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+    });
+
+    let netNodeDevicesRows = netNodeDevices.map(netNodeDevice => {
+        const iface = netNodeDevice.capability.interface;
+        const data = cockpit.format('type=direct,source=$0', iface);
+
+        return (
+            <Select.SelectEntry
+                data={data}
+                key={data}>
```
Comment (martinpitt) [inline pkg/machines/components/create-vm-dialog/pxe-helpers.js]: This could now become one line (totally optional, but as you need to push anyway..)

**Comment by martinpitt** [comment]
@KKoukiou : It seems [ubuntu-stable](https://209.132.184.41:8493/logs/pull-11179-20190312-214040-bc5cc9e2-verify-ubuntu-stable/log.html#249) needs the naughty override as well.

**Comment by KKoukiou** [comment]
@martinpitt this does not have any more testCreate failures anywhere.
https://209.132.184.41:8493/logs/pull-11179-20190313-110404-5dce00a0-verify-fedora-29/log.html#257 cought my attention, looks like a new flake but irrelevant to this PR test. I am currently looking into that.

File: pkg/machines/components/create-vm-dialog/pxe-helpers.js (hunk 20 above)
Comment (KKoukiou) [reply pkg/machines/components/create-vm-dialog/pxe-helpers.js]: Done

File: pkg/machines/components/create-vm-dialog/pxe-helpers.js (hunk 21 above)
Comment (KKoukiou) [reply pkg/machines/components/create-vm-dialog/pxe-helpers.js]: Done

**Comment by martinpitt** [comment]
https://209.132.184.41:8493/logs/pull-11179-20190313-124848-58aa3a2e-verify-rhel-8-0/log.html#257 is the same failure as you pointed out above. As it failed three times in a row, let's make sure it's really unrelated; retrying.

**Comment by martinpitt** [comment]
Meh, it [failed again](https://209.132.184.41:8493/logs/pull-11179-20190313-124848-58aa3a2e-verify-rhel-8-0/log.html#257). This looks real, I'm afraid..

File: pkg/machines/components/nicEdit.jsx [hunk 22]
```diff
@@ -108,7 +108,7 @@ const NetworkTypeAndSourceRow = ({ idPrefix, onValueChanged, dialogValues, netwo
     } else {
         networkSourcesContent = (
             <Select.SelectEntry data='empty' key='empty-list'>
-                <i>{_("No virtual networks")}</i>
+                {_("No virtual networks")}
```
Comment (martinpitt) [inline pkg/machines/components/nicEdit.jsx]: Oh, that was my fault, sorry!
